include/wil/cppwinrt_helpers.h

```diff
@@ -212,5 +212,90 @@ namespace wil::details
     };
 }
 #endif // __WIL_CPPWINRT_MICROSOFT_UI_DISPATCHING_HELPERS
-
 /// @endcond
+
+#if defined(WINRT_Windows_Foundation_Collections_H) && !defined(__WIL_CPPWINRT_WINDOWS_FOUNDATION_COLLECTION_HELPERS)
+#define __WIL_CPPWINRT_WINDOWS_FOUNDATION_COLLECTION_HELPERS
+namespace wil
+{
+    /// @cond
+    namespace details
+    {
+        template<typename T> struct is_vector : std::bool_constant<false> {};
+        template<typename Q> struct is_vector<winrt::Windows::Foundation::Collections::IVector<Q>> : std::bool_constant<true> {};
+        template<typename Q> struct is_vector<winrt::Windows::Foundation::Collections::IVectorView<Q>> : std::bool_constant<true> {};
+        template<typename T> struct is_iterator : std::bool_constant<false> {};
+        template<typename Q> struct is_iterator<winrt::Windows::Foundation::Collections::IIterator<Q>> : std::bool_constant<true> {};

```

Comment:
Should use duck typing so we will work with runtime classes which "quack like" `IVector` or `IVectorView` (`AccessListEntryView`, `DeviceInformationCollection`, `JsonArray`, etc.) without having to fall back to `IIterator`.

Also use duck typing to detect `IIterator` support, so we can support `ResourceMapIterator` and other runtime classes which implement `IIterator`.

```cpp
namespace details
{
    template<typename T>
    struct is_winrt_vector_like {
    private:
        template <typename U,
            typename = decltype(std::declval<U>().GetMany(std::declval<U>().Size(),
                winrt::array_view<decltype(std::declval<U>().GetAt(0))>{}))>
        static constexpr bool get_value(int) { return true; }
        template <typename> static constexpr bool get_value(...) { return false; }  
    public:
        static constexpr bool value = get_value<T>(0);                                                                                                                         
    };

    template<typename T>
    struct is_winrt_iterator_like {
    private:
        template <typename U,
            typename = decltype(std::declval<U>().GetMany(0,
                winrt::array_view<decltype(std::declval<U>().Current())>{}))>
        static constexpr bool get_value(int) { return true; }
        template <typename> static constexpr bool get_value(...) { return false; }  
    public:
        static constexpr bool value = get_value<T>(0);                                                                                                                         
    };
}
```

For `is_winrt_vector_like`, we need `Size()` and `GetMany()`, so I use the expression `o.GetMany(o.Size(), array_view<Elem>{})` where `Elem` is the element type returned by `GetAt()`. Couldn't see an easy way to infer the element type from GetMany, so we end up requiring `GetAt()` too, even though we don't call it.

The `is_winrt_iterator_like` looks just for `GetMany()`, and this time we use `Current()` to infer the element type.

One oddity is that calling a method only from SFINAE sometimes triggers a "Method is declared but never called" warning from MSVC. I think there's an open MSVC bug on it.